#17 Handle cases where type holes are in by name argument closures.

SimY4 opened the pull request:
Handle by name argument closures.

This PR is not covering full-fledged function literals like { a: String => ??? }
I'm currently exploring adding them as well.

SimY4 committed:
#17 Handle cases where type holes are in by name argument closures.

cb372 commented:
Thanks! LGTM.

Do you want to open a separate PR for the other case you mentioned? I'm happy to merge this as-is, or wait for you to add more commits.

SimY4 committed:
#17 Handle cases where type holes are in function literals.
Return type exploration is still something that needs to be figured out.

cb372 committed:
Use parent ValDef's type to get the type of the hole

In the Function tree's type, the `???` is typed as Nothing, which is not
very helpful. So we need to get the type from the parent tree.

cb372 commented:
@SimY4 You were really close! But after a lot of experimenting and `println`-ing, I think the only way to get the type of the hole in this case is to get it from the parent tree.

I've pushed a commit to your branch. Hope you don't mind. I replaced the `case Function(..., Hole(body))` with a `case ValDef(..., Function(..., Hole(body)))` so that we can get the necessary type information from the parent tree's type.

If this looks ok to you, I'll merge this PR and make a release.

EDIT: Forgot to say, thanks again for raising the issue and starting the work. It was really helpful.

SimY4 commented:
@cb372 That's no problem. And thank you for helping with this. I was struggling to find the right way of pattern matching the function case. It's great that you get it right.

SimY4 committed:
#17 support for nested function calls

SimY4 commented:
I added a small change to partially support nested functions. It has a few issues with type resolution. I wonder if it's possible to look up the closes ValDef tree node for any Function to get the real return type.

SimY4 committed:
#17 simplifed example

cb372 commented:
>I wonder if it's possible to look up the closes ValDef tree node for any Function to get the real return type

Unfortunately once you've matched a tree (e.g. a `Function`) it's not possible to walk up the tree from there. So I think our options are:

1. Match on a complex tree like "a ValDef containing a Function", like I did in my commit. But I don't think that approach will really scale to all the different patterns we want to handle.

2. Collect the information we need while we walk down the tree, similar to how we do for bindings. So whenever you traverse a `ValDef` with a function type `A => B` you would store its return type `B` somewhere in case you later encounter a hole in the RHS which should be given its type. In that case I guess the hard part is when you encounter a hole, working out whether its type is given by the return type you stored earlier. If only we could just call `tree.parent` or something, it would be easy...

SimY4 commented:
@cb372 if it's fine with you, can you merge and release these changes, please?

cb372 commented:
Done! Released as 0.0.7.
